test/doc/content/source/userobjects/InterfaceUO_QP.md

```diff
@@ -0,0 +1,16 @@
+# InterfaceUO_QP
+
+## IMPORTANT NOTE
+*** Any derived class of InterfaceUserObject does not support yet getMaterialProperty, please use auxvariables***
+
+## Description
+!syntax description /UserObjects/InterfaceUO_QP
+
+InterfaceUO_QP is a user object computing and storing average values across an interface for each quadrature. The kind of average value are the one available in [InterfaceAverageTools](/InterfaceAverageTools.md).
```

Comment:
Whatever you change the name to for `InterfaceUO` be sure to update this class name. E.g. if `InterfaceUO` becomes `InterfaceAverageUO` then this should become `InterfaceAverageUO_QP`. Which leads me to my primary concern about this PR...we seem to have a ton of objects that are named or do similar things. We have `InterfaceUO` which from its description takes an average, `InterfaceUserObject` which is the base class, and `InterfaceAverageUserObject` which I would think immediately from looking at its name should do something very similar to what `InterfaceUO` is purported to do. When I look at `InterfaceAverageUserObject` I see that most of what it does is to call through to the free-standing functions in `InterfaceAverageTools`. Why can't we have the functionality that's in `InterfaceAverageTools` in the `InterfaceAverageUserObject` class? I'm guessing it's because we also want to use that functionality in the average postprocessor classes... Perhaps we could create an independent object that both the `InterfaceAverageUserObject` and average postprocessor classes inherit from? That would probably be the same amount of code though...so maybe I'm fine with just having that capability in free-standing functions.

So I guess my primary concern is the preponderance of names that appear very similar.

Reply:
@lindsayad thanks for review. I agree that objects' name are a bit confusing and I should something about it. Before making modifications I would like to ask you a question.
As you may have noticed the `InterfaceUO`, `InterfaceUO_QP`, `InterfaceUserObjectQpAux` and `InterfaceUOPPS`, right now are registered in the `MooseTestApp`. is this ok or you prefer to have them available in the framework? If we let them stay the test app we might rename them from `InterfaceUO`->`InterfaceAverageUO_test`, if we put them in the framework we might do as you suggested `InterfaceUO`->`InterfaceAverageUO`.

Also the `InterfaceAverageTools` is a separate functionality because, as you mentioned` we want it to be available in the postprcessor too. Beside saving code with this structure we might add functionality to both the UO and the PP at once, without the risk of making mistake twice.

I also need to fix the `InterfaceIntegralPostprocessor` because as you said it's not computing an integral however this portion of the code is a bit confusing. I really need to understand how the `getValue`, `threadJoin` works, because it seems to me that when calling getValue bypasses `threadJoin`

Reply:
> if we put them in the framework we might do as you suggested `InterfaceUO`->`InterfaceAverageUO`.

I'm in favor of trying to condense objects as long as we are not limiting their general utility. I think the core problem is that for someone who is coming in and looking at this code for the first time, it takes a while to realize that there are two different meanings of "average" here. For someone who is used to the MOOSE system of postprocessors, "average" to them likely means a spatial average over the relevant geometric object (e.g. element or side). But you've introduced a second meaning of "average" here. That second meaning is averaging slave and master values. I think we need to try to clearly delineate these two different meanings.

I think that what the objects `InterfaceUO`, `InterfaceUO_QP`, `InterfaceUserObjectQpAux` do are generally applicable and can be included in the framework. Let's just work on naming and clarity of meaning.

Reply:
@lindsayad if the term average in moose generally associated with a spatial average, In that sense here the only object doing a spatial average is the postprocessor. At an interface we might want an average, a jump, or either something else (basically all the things inside `InterfaceAverageTools`). As you said this is a master slave calculation across the interface, that is not necessarly an average. I need to think about a better name. Let's start from the tools:
what about renaming:

- `InterfaceAverageTolls` -> `InterfaceValueTools`
- `InterfaceUO` -> `InterfaceValueAverageUO` (this is actually computing an average on the Interface)
- `InterfaceUO_QP` -> `InterfaceValueUO_QP`
- `InterfaceUserObjectQpAux` -> `InterfaceValueUO_QP_Aux`

FOR postprocesso the situation is different because they are in fact doing averages of the value. To keep the name short we could do
- `InterfaceIntegralVariableAveragePostprocessor` -> `InterfaceIntegralVariableValuePostprocessor`
- `InterfaceUOPPS` -> `InterfaceValueUOPPS`

waht do you think

Reply:
I like these names much better